_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,359 @@
+
+```
+match may not be exhaustive.
+[warn] It would fail on the following inputs: (NormalSeat(_, _), Standing), (NormalSeat(_, _), VipTribune(_)), (Standing, NormalSeat(_, _)), (Standing, VipTribune(_)), (VipTribune(_), NormalSeat(_, _)), (VipTribune(_), Standing)
+[warn]     (userSeat, guestTicket) match {
+[warn]     ^
+```
+
+The compiler is concerned we might not be considering all possible combinations of seats within a reservation.   
+We are not considering for the cross cases (e.g. (Standing, NormalSeat)) because with the above tricks, we are certain that any created instance of Reservation will always respect the invariant. 
+The compiler isn’t. For it, a **single** Reservation[Seat] might contain tickets in different seating areas. We have to match on both tickets (user and guest) each of which, as far as the compiler goes, can be of any type (<code>Standing</code>, <code>VipTribune</code>, <code>NormalSeated</code>, <code>Seat</code>).  
+Because we (arguably) know more than the compiler in this case, we can tell him not to be concerned with the non-exhaustive pattern match with an annotation (chapter 15.5 at <a id="mmds-ref" href="#programming-in-scala">[1]</a>):   
+
+```scala
+((userSeat, guestTicket): @unchecked) match {
+```
+
+Which will prevent the warning during compilation.
+
+
+# 4. Using OOP - Subtyping
+
+This is the approach most of us would follow. Probably the most pragmatic.
+
+It requires having a bare trait Reservation and extending it for every sub-type of Seat. When there are many sub-types, this would become very repetitive.  
+
+```scala
+sealed trait Reservation[+SeatType <: Seat]
+
+case class ReservationSeated(
+  userTicket: Ticket[NormalSeat],
+  guestTicket: Ticket[NormalSeat]
+) extends Reservation[NormalSeat]
+
+case class ReservationStanding(
+  userTicket: Ticket[Standing.type],
+  guestTicket: Ticket[Standing.type ]
+) extends Reservation[Standing.type]
+
+case class ReservationVipTribune(
+  userTicket: Ticket[VipTribune],
+  guestTicket: Ticket[VipTribune]
+) extends Reservation[VipTribune]
+```
+
+On the other hand, all the type information is carried on at run-time. We can for example pattern match with confidence:  
+
+```scala
+def guideUsersToSeat(reservation: Reservation[Seat]): String = reservation match {
+  case ReservationStanding(_, _) => s"Go to gate 1."
+  case ReservationSeated(userTicket, guestTicket) => s"Go to gate 2. You and your guest are in a seats ${userTicket.seat.column} and ${guestTicket.seat.column} respectively "
+  case ReservationVipTribune(userTicket, _) => s"Go to gate 3. You are on tribune ${userTicket.seat.tribuneId}"
+  }
+```
+
+And because none of our sub-typed Reservation classes has tickets with seats of different types, the invariant is always assured.  
+
+This post is also on the author's blog at []()https://aerodatablog.wordpress.com/
+
+*Do you know any other approach ? I would be happy to hear it. Just write on the comment section.*
```

Comment:
Please remove the extra whitespace before the question mark.

Reply:
done.